File: Easy/191. Number of 1 Bits/solution1.cpp
/*
Author: Djazy Faradj
Problem: 191. Number of 1 Bits
Description: 
 - Given a positive integer n, write a function that returns the number of set bits in its binary representation (also known as the Hamming weight).

Note: This problem thought me two nice concepts:    1. The existance of bitwise operators, to perform bitwise AND (&) and bitwise OR (|)
                                                    2. The meaning of "Hamming weight" which is the number of set bits in a binary number.
      This solution beats 100% in terms of runtime and 76% in terms of memory.

Time Complexity: O(n) where n is the number of bits in the binary representation of the number
Space Complexity: O(1)

Relevent Topics: Divide and conquer, Bit Manipulation
*/

class Solution {
public:
    int hammingWeight(int n) {
        uint8_t count = 0;
        while (n)
        {
            n &= (n-1);
            count ++;
        }
        return count;
    }
};

// 1011     n       (11)
// 1010     n-1     (10)

// 1010     n       (10)    count = 1
// 1001     n-1     (9)     

// 1000     n       (8)     count = 2
// 0111     n-1     (7) 

// 0000     n       (0)     count = 3

// STOPS

File: Easy/226. Invert Binary Tree/solution1.java
/*
Author: Djazy Faradj
Problem: 226. Invert Binary Tree
Description: 
 - Given the root of a binary tree, invert the tree, and return its root.

Note: This is the solution I provided to my Data Structure and Algorithms class question.
Time Complexity: O(i) where i represents number of internal nodes
Space Complexity: ?

Relevent Topics: Tree, DFS, BFS, Binary Tree
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isExternal(TreeNode v) {
        if (v == null) return false;
        return (v.left == null && v.right == null);
    }

    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;

        if (!isExternal(root.left)) {
            invertTree(root.left);
        }

        if (!isExternal(root.right)) {
            invertTree(root.right);
        }

        TreeNode temp = root.left == null ? null : new TreeNode(root.left.val, root.left.left, root.left.right);
        root.left = root.right;
        root.right = temp;
        return root;
    }
}